code/modules/reagents/Chemistry-Reagents/Drug-Reagents.dm

```diff
@@ -355,10 +352,95 @@
 			step(M, pick(cardinal))
 	M.Jitter(50)
 	M.Dizzy(50)
-	M.adjustToxLoss(5)
+	M.adjustToxLoss(4*REM)
 	M.adjustBrainLoss(10)
 	if(prob(50))
-		M.emote(pick("twitch","drool","moan"))
+		M.emote(pick("twitch","drool","moan","vomit","flip","scream"))
+	..()
+	return
+
+/datum/reagent/drug/heroin
+	name = "Heroin"
+	id = "heroin"
+	description = "An extremely advanced painkiller/narcotic. Heroin allows you to ignore all slowdown and grants you full immunity to stamina damage, but stuns are twice as effective against you. Mildly toxic."
+	reagent_state = LIQUID
+	color = "#C8A5DC"
+	metabolization_rate = 0.75 * REAGENTS_METABOLISM
+	overdose_threshold = 15
+	addiction_threshold = 10
+	speedboost = IGNORE_SLOWDOWN
+
+/datum/reagent/drug/heroin/on_mob_life(mob/living/M)
+	M.setStaminaLoss(0)
+	if(M.stunned || M.weakened || M.paralysis)
+		M.AdjustStunned(0.5)
+		M.AdjustWeakened(0.5)
+	if(iscarbon(M))
+		var/mob/living/carbon/N = M
+		N.hal_screwyhud = 5
+	M.adjustToxLoss(0.3*REM)
+	M.adjustFireLoss(-0.5*REM)
+	M.adjustBruteLoss(-0.5*REM)
+	..()
+	return
+
+/datum/reagent/drug/heroin/on_mob_delete(mob/living/M)
+	if(iscarbon(M))
+		var/mob/living/carbon/N = M
+		N.hal_screwyhud = 0
+	..()
+
+/datum/reagent/drug/heroin/overdose_process(mob/living/M)
+	if(prob(33))
+		var/obj/item/I = M.get_active_hand()
+		if(I)
+			M.drop_item()
+		M.Dizzy(2)
+		M.Jitter(2)
+	if(prob(10))
```

Comment:
This like a 1% or less chance to end up with >7 sleeping.
It'd be nice if it checked you weren't already asleep though.